### PR TITLE
Attempt at making playground useful + tweaks with reversing display of grid

### PR DESCRIPTION
**Pull request by bahmady**
I've attempted to make playground useful to make a kind of CLI game loop, using `loop`:

```
$ cat playground.rb
require './lib/minesweeper'
game = Minesweeper::Game.new(5, 5)

loop do
  x = gets.chomp.to_i - 1
  y = gets.chomp.to_i - 1
  puts game.unseal(x, y)
  break if game.status == :game_lost
end
```

I use `-1` in the user input for the two coordinates because I want it to feel like 0,0 is 1,1 etc.

To do this I made `Game::@status` readable:

`attr_reader :status`

I have also used `reverse` in `Game::display` twice to make it display the grid as I expect (where 0,0 is bottom left and 4,4 is top right):

```
    def display
      @grid.state.map do |row|
        row_display = row.map{ |cell| cell.nil? ? " . " : " #{cell} " }.join("")
        "#{row_display}\n".reverse
      end.join("").reverse
    end
```

My question would be (beyond whether the code is orthodox) whether the responsibility here is fair, i.e. using playground this way, rather than making new methods of some sort and putting them in `Game` (and renaming it `Game_CLI` or something). 

It feels like this is sensible, because I've effectively made something out of the "library" of Minesweeper which we created. And that something happens to be a purely CLI game. But then the way you've used `#display` makes me wonder.

### CODE REVIEW

**Comment by arnau** [comment]
> I've attempted to make playground useful to make a kind of CLI game loop, using `loop`:
> 
> ```
> $ cat playground.rb
> require './lib/minesweeper'
> game = Minesweeper::Game.new(5, 5)
> 
> loop do
>   x = gets.chomp.to_i - 1
>   y = gets.chomp.to_i - 1
>   puts game.unseal(x, y)
>   break if game.status == :game_lost
> end
> ```
> 
> I use `-1` in the user input for the two coordinates because I want it to feel like 0,0 is 1,1 etc.

Ok, if it is something that you want to have in this interface only, it's a way. If you want to have _everyone_ and _everything_ to use indexes starting at (1,1) then move this logic to `Game`.

> 
> To do this I made `Game::@status` readable:
> 
> `attr_reader :status`

This is the opportunity to use the `#lost?` method. Again, leaking is rarely the answer. In this case, leaking symbols is not a good idea.

```ruby
break if game.lost?
```

The above is more robust.

> 
> I have also used `reverse` in `Game::display` twice to make it display the grid as I expect (where 0,0 is bottom left and 4,4 is top right):

Interesting, (0, 0) in screens starts at (top, left) but sure, if you expect it in another way change it. Why do you reverse each row though? Doesn't this code makes (0, 0) the (bottom, right) corner?

Importantly, `#reverse` should be applied before `#map`. Doing it after may give you unexpected results because you are reversing strings instead of arrays.

> 
> ```
>     def display
>       @grid.state.map do |row|
>         row_display = row.map{ |cell| cell.nil? ? " . " : " #{cell} " }.join("")
>         "#{row_display}\n".reverse
>       end.join("").reverse
>     end
> ```
> 
> My question would be (beyond whether the code is orthodox) whether the responsibility here is fair, i.e. using playground this way, rather than making new methods of some sort and putting them in `Game` (and renaming it `Game_CLI` or something).

Playground is acting as the CLI, `Game` is not. That said, there is an opportunity to decouple game from the text-based UI, but depending on your goal it might be overkill.

> 
> It feels like this is sensible, because I've effectively made something out of the "library" of Minesweeper which we created. And that something happens to be a purely CLI game. But then the way you've used `#display` makes me wonder.

Why is that?

**Comment by bahmady** [comment]
> Ok, if it is something that you want to have in this interface only, it's a way. If you want to have _everyone_ and _everything_ to use indexes starting at (1,1) then move this logic to `Game`.

I feel like it makes sense to just have it in `playground` for the time being, because implementations using this code (hypothetically) might expect it to follow standard Ruby array indexing. Is that a load of nonsense? It makes sense to me. 

> This is the opportunity to use the `#lost?` method. Again, leaking is rarely the answer. In this case, leaking symbols is not a good idea.

OK, that makes a lot of sense. I'll try to avoid leaking in general and I've created `Game::#lost`. I actually still had some leaking for `@rows` and `@columns` in `Grid` and I don't know why (I shouldn't have because now that I think about it you told me they shouldn't be there some time ago... sorry). The code functions fine without them. 

> Interesting, (0, 0) in screens starts at (top, left) but sure, if you expect it in another way change it. Why do you reverse each row though? Doesn't this code makes (0, 0) the (bottom, right) corner?

The code I _had_ actually did output as I said, even with reversing each row. I'm not sure why either. But ... [cont] 👇 

> Importantly, `#reverse` should be applied before `#map`. Doing it after may give you unexpected results because you are reversing strings instead of arrays.

...  [cont]☝️ I did this and now I only need one `#reverse`. 

> Playground is acting as the CLI, `Game` is not. That said, there is an opportunity to decouple game from the text-based UI, but depending on your goal it might be overkill.
> 
> > It feels like this is sensible, because I've effectively made something out of the "library" of Minesweeper which we created. And that something happens to be a purely CLI game. But then the way you've used `#display` makes me wonder.
> 
> Why is that?

Answering this in one block. It makes sense that `playground` is acting as the CLI, but `Game` is limited by outputting as a text-based UI. I'm conflating the two. Having thought about it, I'd like to explore decoupling `Game` and see if I can use some sort of PNG library or something next to render the game that way. That's more of an exercise I'd like to explore; I don't know if I'll be successful. For that reason the approach in this PR makes sense to me. 

To put it another way: it makes sense to me for the long term, given my goals to use `playground` this way (as in, in this PR) rather than to hard-code some sort of method in `Game` called e.g. `loop`. Given that `Game` is limited by what it can display (text-based UI, from `#display`) it seems like if I _do not_ decouple `Game`, it would actually make sense to hard-code the e.g. `loop` method into `Game` and call this whole thing a finished project there named "Minesweeper CLI" or somesuch. 

Sorry if that explanation is a bit poor.

**Comment by arnau** [comment]
If you attempt to add a PNG-based UI, you'll face with the situation of having to reuse code in Game that now is specific to text-based UI. That means you'll probably need to extract that and keep Game UI agnostic. If you don't, and say merge playground into Game, you'll have to create another fully fleshed class that implements the Game logic, with the difference of how it outputs. If it's unclear why this is a problem, just do it. If it seems fine still, we can discuss it over the actual implementation.

It's important that you know why an abstraction exists so it is important that you experience it yourself.
